projects/Test/Integration/TestBasicPublish.cs

```diff
@@ -322,5 +322,73 @@ public async Task TestPropertiesRoundtrip_Headers()
             Assert.Equal(sendBody, consumeBody);
             Assert.Equal("World", response);
         }
+
+        [Fact]
+        public async Task TestPublisherConfirmationThrottling()
+        {
+            const int MaxOutstandingConfirms = 4;
+
+            var channelOpts = new CreateChannelOptions
+            {
+                PublisherConfirmationsEnabled = true,
+                PublisherConfirmationTrackingEnabled = true,
+                OutstandingPublisherConfirmationsRateLimiter = new ThrottlingRateLimiter(MaxOutstandingConfirms)
+            };
+
+            var channelCreatedTcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var messagesPublishedTcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+            Task publishTask = Task.Run(async () =>
+            {
+                ConnectionFactory cf = CreateConnectionFactory();
+
+                await using (IConnection conn = await cf.CreateConnectionAsync())
+                {
+                    await using (IChannel ch = await conn.CreateChannelAsync(channelOpts))
+                    {
+                        QueueDeclareOk q = await ch.QueueDeclareAsync();
+
+                        channelCreatedTcs.SetResult(true);
+
+                        int publishCount = 0;
+                        /*
+                         * Note: if batchSize equals MaxOutstandingConfirms,
+                         * a delay is added per-publish and this test takes much longer
+                         * to run. TODO figure out how the heck to test that
```

Comment:
Newer version of .NET have a timeprovider abstraction https://andrewlock.net/exploring-the-dotnet-8-preview-avoiding-flaky-tests-with-timeprovider-and-itimer/

There is also a .NET Standard implementation for that https://www.nuget.org/packages/Microsoft.Bcl.TimeProvider/#readme-body-tab but that is a whole other set of things you might want to avoid at this stage?

Reply:
I think the best option right now is to set up a toxiproxy test that blocks responses from RabbitMQ, which will block confirmations. Going to give that a try later today.